python/cudf/cudf/_lib/pylibcudf/io/csv.pyx

```diff
@@ -0,0 +1,292 @@
+        for col in parse_dates:
+            if isinstance(col, str):
+                c_parse_dates_names.push_back(col.encode())
+            elif isinstance(col, int):
+                c_parse_dates_indexes.push_back(col)
+            else:
+                raise NotImplementedError(
+                    "`parse_dates`: Must pass a list of column names/indices")
+
+        # Set both since users are allowed to mix column names and indices
+        options.set_parse_dates(c_parse_dates_names)
+        options.set_parse_dates(c_parse_dates_indexes)
+
+    if parse_hex is not None:
+        for col in parse_hex:
+            if isinstance(col, str):
+                c_parse_hex_names.push_back(col.encode())
+            elif isinstance(col, int):
+                c_parse_hex_indexes.push_back(col)
+            else:
+                raise NotImplementedError(
+                    "`parse_hex`: Must pass a list of column names/indices")
+        # Set both since users are allowed to mix column names and indices
+        options.set_parse_hex(c_parse_hex_names)
+        options.set_parse_hex(c_parse_hex_indexes)
+
+    cdef string k_str
+    if isinstance(dtypes, list):
+        for dtype in dtypes:
+            if not isinstance(dtype, DataType):
+                raise TypeError("If passing list to read_csv, "
+                                "all elements must be of type `DataType`!")
+            c_dtypes_list.push_back((<DataType>dtype).c_obj)
+        options.set_dtypes(c_dtypes_list)
+    elif isinstance(dtypes, dict):
+        # dtypes_t is dict
+        for k, v in dtypes.items():
+            k_str = str(k).encode()
+            if not isinstance(v, DataType):
+                raise TypeError("If passing dict to read_csv, "
+                                "all values must be of type `DataType`!")
+            c_dtypes_map[k_str] = (<DataType>v).c_obj
+        options.set_dtypes(c_dtypes_map)
+    elif dtypes is not None:
+        raise TypeError("dtypes must either by a list/dict")
+
+    if true_values is not None:
+        c_true_values.reserve(len(true_values))
+        for tv in true_values:
+            if not isinstance(tv, str):
+                raise TypeError("true_values must be a list of str!")
+            c_true_values.push_back(tv.encode())
+        options.set_true_values(c_true_values)
+
+    if false_values is not None:
+        c_false_values.reserve(len(false_values))
+        for fv in false_values:
+            if not isinstance(fv, str):
+                raise TypeError("false_values must be a list of str!")
+            c_false_values.push_back(fv.encode())
+        options.set_false_values(c_false_values)
+
+    if na_values is not None:
+        c_na_values.reserve(len(na_values))
+        for nv in na_values:
+            if not isinstance(nv, str):
+                raise TypeError("na_values must be a list of str!")
+            c_na_values.push_back(nv.encode())
+        options.set_na_values(c_na_values)
```

Comment:
This type of code is repeated a lot in this function's parsing of inputs. It feels like a helper function along the lines of
```
cdef vector[T] vec_from_iterable(vec, test, error_msg)
```
could help. OTOH maybe that's overengineering since you'll still have to write predicate functions and error messages each time. Maybe give it a shot once and see what it looks like.

Reply:
Kinda did something like that.